python/ray/util/client/dataclient.py

```diff
@@ -73,66 +78,157 @@ def _start_datathread(self) -> threading.Thread:
+                metadata = self._metadata + \
+                    [("reconnecting", str(reconnecting))]
+                resp_stream = stub.Datapath(
+                    iter(self.request_queue.get, None),
+                    metadata=metadata,
+                    wait_for_ready=True)
+                try:
+                    for response in resp_stream:
+                        self._process_response(response)
+                    return
+                except grpc.RpcError as e:
+                    reconnecting = self._can_reconnect(e)
+                    if not reconnecting:
+                        return
+                    self._reconnect_channel()
+        except Exception as e:
+            self._last_exception = e
+        finally:
+            logger.info("Shutting down data channel")
+            self._shutdown()
+
+    def _process_response(self, response: Any) -> None:
+        """
+        Process responses from the data servicer.
+        """
+        if response.req_id == 0:
+            # This is not being waited for.
+            logger.debug(f"Got unawaited response {response}")
+            return
+        if response.req_id in self.asyncio_waiting_data:
+            try:
+                # NOTE: calling self.asyncio_waiting_data.pop() results
+                # in the destructor of ClientObjectRef running, which
+                # calls ReleaseObject(). So self.asyncio_waiting_data
+                # is accessed without holding self.lock. Holding the
+                # lock shouldn't be necessary either.
+                callback = self.asyncio_waiting_data.pop(response.req_id)
+                if callback:
+                    callback(response)
+            except Exception:
+                logger.exception("Callback error:")
             with self.lock:
-                self._in_shutdown = True
-                self._last_exception = e
+                # Update outstanding requests
+                if response.req_id in self.outstanding_requests:
+                    del self.outstanding_requests[response.req_id]
+                    # Acknowledge response
+                    self._acknowledge(response.req_id)
+        else:
+            # Blocking response. Populate ready data and wake up waiting
+            # threads.
+            with self.cv:
+                self.ready_data[response.req_id] = response
                 self.cv.notify_all()
 
-                callbacks = self.asyncio_waiting_data.values()
-                self.asyncio_waiting_data = {}
+    def _can_reconnect(self, e: grpc.RpcError) -> bool:
+        """
+        Processes RPC errors that occur while reading from data stream.
+        Returns True if the error can be recovered from, False otherwise.
+        """
+        if not self.client_worker._can_reconnect(e):
+            logger.info("Unrecoverable error in data channel.")
+            logger.debug(e)
+            self._last_exception = e
```

Comment:
I personally would find it easier to read, to set `self._last_exception` inside the `if not reconnecting:` condition.